kratos/includes/kratos_parameters.h

```diff
@@ -54,969 +51,1541 @@ namespace Kratos {
-    using value_iterator = rapidjson::Value::ConstMemberIterator;
-    value_iterator mValueIterator;
-    std::unique_ptr<Parameters> mpParameters;
-
-  public:
-    const_iterator_adaptor(value_iterator it,
-                           Kratos::shared_ptr<rapidjson::Document> pdoc)
-        : mValueIterator(it),
-          mpParameters(new Parameters(
-              const_cast<rapidjson::Value *>(&it->value), pdoc)) {}
-    const_iterator_adaptor(const const_iterator_adaptor &it)
-        : mValueIterator(it.mValueIterator),
-          mpParameters(new Parameters(*(it.mpParameters))) {}
-    const_iterator_adaptor &operator++() {
-      mValueIterator++;
-      return *this;
-    }
-    const_iterator_adaptor operator++(int) {
-      const_iterator_adaptor tmp(*this);
-      operator++();
-      return tmp;
-    }
-    bool operator==(const const_iterator_adaptor &rhs) const {
-      return mValueIterator == rhs.mValueIterator;
-    }
-    bool operator!=(const const_iterator_adaptor &rhs) const {
-      return mValueIterator != rhs.mValueIterator;
-    }
-    const Parameters &operator*() const {
-      mpParameters->SetUnderlyingSotrage(
-          const_cast<rapidjson::Value *>(&mValueIterator->value));
-      return *mpParameters;
-    }
-    const Parameters *operator->() const {
-      mpParameters->SetUnderlyingSotrage(
-          const_cast<rapidjson::Value *>(&mValueIterator->value));
-      return mpParameters.get();
-    }
-    value_iterator &base() { return mValueIterator; }
-    value_iterator const &base() const { return mValueIterator; }
-    std::string name() { return mValueIterator->name.GetString(); }
-  };
-
-  ///@}
+    ///@name Nested clases
+    ///@{
+    /**
+     * @class iterator_adaptor
+     * @ingroup KratosCore
+     * @brief This nested class can be used to adapt a Parameter iterator
+     * @author Riccardo Rossi
+     */
+    class iterator_adaptor
+        : public std::iterator<std::forward_iterator_tag, Parameters>
+    {
+        ///@name Type Definitions
+        ///@{
+
+        using value_iterator = nlohmann::json::iterator; /// Iterator definition
+
+        ///@}
+        ///@name Member Variables
+        ///@{
+
+        value_iterator mValueIterator;                   /// Our iterator
+        nlohmann::json& mrValue;                         /// The original container
+        std::unique_ptr<Parameters> mpParameters;        /// The unique pointer to the base Parameter
+
+        ///@}
+    public:
+        ///@name Life Cycle
+        ///@{
+
+        /**
+         * @brief Default constructor (iterator + root Parameter)
+         * @param itValue The iterator to adapt
+         * @param pRoot The root Parameter pointer
+         */
+        iterator_adaptor(value_iterator itValue, nlohmann::json* pValue,  Kratos::shared_ptr<nlohmann::json> pRoot) :mValueIterator(itValue), mrValue(*pValue), mpParameters(new Parameters(itValue, pValue, pRoot)) {}
```

Comment:
I remember you had a problem with the iterators. how is that solved?

also, is current version valgrind-clean?

Reply:
It is valgring clean

Reply:
The problem was solved after asking to json developers, and the recommended to save the reference of the original container and check that the iterator to dereference is not an end() iterator